### DISCUSSION TITLE
#22 ESP32 Support

### DISCUSSION THREAD

**Discussion by lkankowski** [lkankowski/arduino-multi-relay2#26] [Q&A]
Przenjesiona dyskusja z Issue #22 
@pepsonEL 
> Na jakiej zasadzie chcesz wprowadzić obsługę ESP32 ? Zamiast Arduino ?

> @lkankowski lkankowski commented 13 minutes ago

> Tak samo jak z ESP8266 - po prostu kwestia dodania konfiguracji kompilacji, bo MySensors ma już obsługę. No i pewnie kilka drobnych zmian, żeby się kompilowało.

> @pepsonEL pepsonEL commented 11 minutes ago
> Czyli ESP w połączeniu z Arduino tak ?

> @lkankowski lkankowski commented 8 minutes ago
> Nie - w całości na ESP. Nazwa Arduino odnosi się nie tylko do płytek, ale też do frameworka. ESP także można programować we frameworku Arduino :)

> @pepsonEL pepsonEL commented 1 minute ago
> Czyli że soft wgrywamy na ESP32 i do niego np trzeba podłaczyć ekspandery aby obsługiwać jak do tej pory przekaźniki i przełaczniki ścienne ? Bo teraz część podłaczeń szło na arduino mega a reszta jak coś przez ekspandery....

**Comment by lkankowski** [comment]
@pepsonEL Zgadza się. To tak jakby zamienić Mega na ESP32.
Istnieje oczywiście opcja, że ESP służy jedynie do komunikacji Wifi i w innym wątku ostatnio drążyliśmy ten temat. Są nawet płytki w formacie Arduino Mega, na której także jest ESP8266. Tylko okazuje się, że to się programuje jak 2 osobne płytki, więc na Mege ładowałoby się standardowy sketch z komunikacją UART, a na ESP trzeba załadować coś innego, co przekazuje UART przez sieć lub od razu do MQTT (np. EspEasy).

**Comment by pepsonEL** [comment]
No to powiem Ci super jakby wprowadzić ESP32 może wtedy szybciej by to działało. Chyba lepsze jest os Arduino.... To jak w moim wypadku może by sprawnie działało wtedy gaszenie wszystkich świateł na raz.....

**Comment by lkankowski** [comment]
U Ciebie problemem może być MCP23017 - one są dość wolne, co nie jest problemem przy przekaźnikach, ale już dla przełączników z funkcjami double/long click robi różnice. Ostatnio natknąłem się na szybszą wersję - MCP23S17.
Chociaż jeżeli chodzi o masowe gaszenie w HA, to problem może być zupełnie inny - przepustowość UART.
Może podniesienie prędkości powyżej 115k poprawiłoby coś? Np. ESP8266 spokojnie daje rade z 460k, a nawet 920k :)

**Comment by lkankowski** [comment]
Aha - przejście na natywne MQTT też mogłoby rozwiązać problem, bo wtedy broker kolejkuje komunikaty.
Podstawowa obsługa już jest w oparciu o MySensors i nawet dla Mega z W5100, ale podobno MySensors może mieć problem z reconnect i będę to sprawdzał.
Myślałem też, żeby zrobić wariant bez MySensors - bezpośrednio MQTT Auto Discovery.

**Reply by pepsonEL** [reply]
I to byłoby fajne.... aby wtedy HA wykryło samo urządzenia.

A co do tego ze mam ekspander. Wcześniej kiedyś miałem tylko dwa arduino bez ekspanderów i od zawsze był problem z tym gaszeniem.

**Comment by lkankowski** [comment]
Wydaje się, że problemem jest zbyt duży strumień danych w UART, a może i jakiś błąd w obsłudze.
Dla mnie nie ma problemu z dodaniem dodatkowego polecenia, które wyłącza wszystko, tylko czy HA to obsłuży?
MySensors ma np takie komendy: V_SCENE_ON, V_SCENE_OFF (urządzenie S_SCENE_CONTROLLER) lub V_CUSTOM.
Przy V_SCENE_OFF musiałbym założyć, że wyłączam wszystkie przekaźniki (albo z wyjątkiem tych, co mają STARTUP_ON/OFF).
Pytanie, czy HA to obsługuje?

**Reply by pepsonEL** [reply]
tego nie wiem....

**Comment by lkankowski** [comment]
https://www.home-assistant.io/integrations/sensor.mysensors/
Z tego wynika, że obie opcje są obsługiwane.
Pytanie do Ciebie, czy wyłączanie wszystkiego jest OK, czy może są wyjątki?

**Comment by pepsonEL** [comment]
Sa wyjatki bo 4 przekaxniki sa od bram i jeden od kontaktronu.....

**Comment by lkankowski** [comment]
A czymś się wyróżniają w konfiguracji, tak, aby można je wykluczyć?

**Reply by pepsonEL** [reply]
Te od bram to np maja impuls.....

**Comment by lkankowski** [comment]
No to mógłbym zrobić wyłączanie wszystkich z wyjątkiem:
- RELAY_STARTUP_ON
- RELAY_STARTUP_OFF
- RELAY_IMPULSE

**Reply by pepsonEL** [reply]
jeszcze też wtedy od kontaktronów.....

**Comment by pepsonEL** [comment]
Myślę że lepiej jak by to działało że w HA się robi grupę all i tam user sam wrzuca sobie które mają się gasić....

**Comment by lkankowski** [comment]
No i się komplikuje :)
Też myślę, że grupy/sceny lepiej robić w systemie Home Automation, to tylko taka luźna sugestia workarounda.

**Comment by szurr** [comment]
Czytam i się zastanawiam jaki sens mnożyć problemy z mysensors pod esp32 dla HA. 
Przecież esphome działa dużo lepiej jeśli chodzi o zgodność grup typów itp a do tego masz ota masz esp z ethernetem i konfiguracja to chyba 5 linijek. Może warto rozważyć

**Reply by pepsonEL** [reply]
No tak jak ja mam ok 60 przekaźników to nie takie 5 linijek ale owszem ESPHome działa lepiej to prawda...

**Reply by lkankowski** [reply]
A co konkretnie działa lepiej?

**Reply by pepsonEL** [reply]
Szybkość i kompatybilność z HA. Że światła są jako light a nie switch... Opcja wgrywania OTA...

**Reply by lkankowski** [reply]
A już myślałem, że coś nie działa, a Ty piszesz o ficzerach :P

**Reply by lkankowski** [reply]
@pepsonEL A jeszcze dopytam o szybkość - jak to mierzysz?

**Reply by pepsonEL** [reply]
Szybkość miałem na myśli to gaszenie wszystkich przekaźników jednym kliknięciem.

**Comment by lkankowski** [comment]
Dodanie ESP32 jak już jest ESP8266 to żaden wysiłek :)
Możliwe, że to i tak będzie tylko przejściowe rozwiązanie przed MQTT Auto Discovery - przynajmniej dla wersji innych niż serial.

**Reply by szurr** [reply]
No brakuje mi trochę seriala w esphome ale za to cała reszta dość mocno idzie do przodu.
